daemon/container.go

```diff
@@ -291,6 +291,21 @@ func (daemon *Daemon) verifyContainerSettings(platform string, hostConfig *conta
 		return nil, nil
 	}
 
+	if hostConfig.Privileged {
+		for _, deviceMapping := range hostConfig.Devices {
+			if deviceMapping.PathOnHost == deviceMapping.PathInContainer {
+				continue
+			}
+			if _, err := os.Stat(deviceMapping.PathInContainer); err != nil {
+				if os.IsNotExist(err) {
+					continue
+				}
+				return nil, errors.Wrap(err, "error stating device path in contianer")
+			}
+			return nil, errors.Errorf("container device path: %s must be different from any host device path for privileged mode containers", deviceMapping.PathInContainer)
```

Comment:
In what situation do we end here? (looks like we only get here if `deviceMapping.PathOnHost != deviceMapping.PathInContainer` and if `deviceMapping.PathInContainer` exists.

But the _error_ says (if I interpret correctly); it's not allowed to add a device if `deviceMapping.PathOnHost == deviceMapping.PathInContainer` ?

Reply:
Code changed to make it more clear

Reply:
@thaJeztah `deviceMapping.PathOnHost == deviceMapping.PathInContainer` has a `continue`

Reply:
Trying to go through the logic:

1. `--device /dev/vdc:/dev/vdc`, means: `deviceMapping.PathOnHost == deviceMapping.PathInContainer`
  - continue (no error)
2. `--device /dev/vdc:/some/path`, and `/some/path` doesn't exist in the container:
  - continue (no error)
3. `--device /dev/vdc:/some/path`, and `/some/path` cannot be read in the container:
  - error: `error stating device path in container`
4. `--device /dev/vdc:/some/path`, and _exists_ in the container:
  - error: `container device path: /some/path must be different from any host device path for privileged mode containers`

So for 4., the device target in the container is not allowed to overlap with an existing path in the container;

- Is this different for non-privileged containers? (I _guess_ not; but we don't seem to be validating that)
- If `deviceMapping.PathOnHost == deviceMapping.PathInContainer`, should we still validate if `PathInContainer` is not an existing path in the container? (could there be situations where a device exists on the host at `/foo/bar`, but the container has a directory (or file) at that location?)
- The error message confuses me, because (IIUC) it's actually about the _container_ path (not the _host_ path), correct?

(again, apologies if I'm misunderstanding :smile:)

Reply:
@thaJeztah 
For step 4:
The device target in the container is not allowed to overlap with an existing path on the host but not in the container.
For privileged mode containers , some hostpath maybe mapped into the container, so I do this check here. I admit it is a bit excessive, or we just need to check the `/dev/` path?  What do you think?